src/Components/Web.JS/src/Rendering/JSRootComponents.ts

```diff
@@ -0,0 +1,84 @@
+import { DotNet } from '@microsoft/dotnet-js-interop';
+
+const pendingRootComponentContainerNamePrefix = '__bl-dynamic-root:';
+const pendingRootComponentContainers = new Map<string, Element>();
+let nextPendingDynamicRootComponentIdentifier = 0;
+
+type ComponentParameters = object | null | undefined;
+const textEncoder = new TextEncoder();
+
+let manager: DotNet.DotNetObject | undefined;
+
+// These are the public APIs at Blazor.rootComponents.*
+export const RootComponentsFunctions = {
+    async add(toElement: Element, componentIdentifier: FunctionStringCallback, initialParameters: ComponentParameters): Promise<DynamicRootComponent> {
+        if (!initialParameters) {
+            throw new Error('initialParameters must be an object, even if empty.');
+        }
+
+        // Track the container so we can use it when the component gets attached to the document via a selector
+        const containerIdentifier = pendingRootComponentContainerNamePrefix + (++nextPendingDynamicRootComponentIdentifier).toString();
+        pendingRootComponentContainers.set(containerIdentifier, toElement);
+
+        // Instruct .NET to add and render the new root component
+        const componentId = await getRequiredManager().invokeMethodAsync<number>(
+            'AddRootComponent', componentIdentifier, containerIdentifier);
+        const component = new DynamicRootComponent(componentId);
+        await component.setParameters(initialParameters);
+        return component;
+    }
+};
+
+export function getAndRemovePendingRootComponentContainer(containerIdentifier: string): Element | undefined {
+    const container = pendingRootComponentContainers.get(containerIdentifier);
+    if (container) {
+        pendingRootComponentContainers.delete(containerIdentifier);
+        return container;
+    }
+}
+
+class DynamicRootComponent {
+    private _componentId: number | null;
+
+    constructor(componentId: number) {
+        this._componentId = componentId;
+    }
+
+    setParameters(parameters: ComponentParameters) {
+        parameters = parameters || {};
+        const parameterCount = Object.keys(parameters).length;
+
+        // TODO: Need to use the JSInterop serializer here so that special objects like JSObjectReference
```

Comment:
Are you planning to tackle this next?

Reply:
It's the same issue as https://github.com/dotnet/aspnetcore/issues/34338 so will be fixed as part of that (during this milestone).